### PR TITLE
Use CSRF protection if available

### PR DESCRIPTION
**Pull request by rnewson**
Use CSRF protection if available

COUCHDB-2762

### CODE REVIEW

**Comment by rnewson** [comment]
Just for demonstration purposes, I expect there's better ways to do this.

File: app/app.js [hunk 1]
```diff
@@ -63,6 +63,29 @@ function (app, $, _, Backbone, Bootstrap, Helpers, Utils, FauxtonAPI, Couchdb) {
   // Localize or create a new JavaScript Template object
   var JST = window.JST = window.JST || {};
 
+  var parseCookies = function (cookies) {
```
Comment (garrensmith) [inline app/app.js]: You can simplify this function by using [`_.reduce`](http://underscorejs.org/#reduce).

File: app/app.js [hunk 2]
```diff
@@ -63,6 +63,29 @@ function (app, $, _, Backbone, Bootstrap, Helpers, Utils, FauxtonAPI, Couchdb) {
...
+    var list = {}, rc = cookies;
```
Comment (garrensmith) [inline app/app.js]: Can you put each argument declaration on a new line.

File: app/app.js [hunk 3]
```diff
@@ -63,6 +63,29 @@ function (app, $, _, Backbone, Bootstrap, Helpers, Utils, FauxtonAPI, Couchdb) {
...
+  };
+
+  $._ajax = $.ajax;
+
+  $.ajax = function (settings) {
+    var cookies = parseCookies(document.cookie);
+    var csrf = cookies['CouchDB-CSRF'] ? cookies['CouchDB-CSRF'] : 'true';
+    var origBeforeSend = settings.beforeSend;
+    var newBeforeSend = function (xhr) {
+      if (origBeforeSend) origBeforeSend(xhr);
```
Comment (garrensmith) [inline app/app.js]: Can you add `{` and `}` for the if statement.

File: app/app.js [hunk 4]
```diff
@@ -63,6 +63,29 @@ function (app, $, _, Backbone, Bootstrap, Helpers, Utils, FauxtonAPI, Couchdb) {
   // Localize or create a new JavaScript Template object
   var JST = window.JST = window.JST || {};
 
+  var parseCookies = function (cookies) {
+    var list = {}, rc = cookies;
+    rc && rc.split(';').forEach(function (cookie) {
```
Comment (garrensmith) [inline app/app.js]: Can you rather change the `rc` check to an if statement. Its better to make it a bit more explicit.

File: app/app.js [hunk 5]
```diff
@@ -63,6 +63,29 @@ function (app, $, _, Backbone, Bootstrap, Helpers, Utils, FauxtonAPI, Couchdb) {
...
+      var parts = cookie.split('=');
+      list[parts.shift().trim()] = decodeURI(parts.join('='));
+    });
+    return list;
+  };
+
+  $._ajax = $.ajax;
+
+  $.ajax = function (settings) {
```
Comment (garrensmith) [inline app/app.js]: Instead of overriding `$.ajax` could we rather use a [global setting](https://api.jquery.com/jQuery.ajaxSetup/)?

Comment (rnewson) [inline app/app.js]: we'll need to derive the right value from the current cookie (or absence of), it's not a constant.

**Comment by rnewson** [comment]
@garren are we good to go now?

**Comment by garrensmith** [comment]
I don't love the fact we overriding ajax. But I cannot think of a better way. So +1

**Comment by rnewson** [comment]
An iteration over this would be for all our currently $.ajax calls to call a function of our own making. I did it this way to be sure I got them all.

**Pull request merged**